### ISSUE TITLE
SequelizeConnectionAcquireTimeoutError randomly DB connections pool error

### ISSUE THREAD

**Issue by idangozlan** [sequelize/sequelize#10858]
I'm running a very simple app that using sequelize basics, it was working for the recent year without any issues on sequelize v4.
Recently, the app is crashing every day once a day, with the following error:

```
TimeoutError: ResourceRequest timed out
    at ResourceRequest._fireTimeout (/usr/src/app/node_modules/generic-pool/lib/ResourceRequest.js:62:17)
    at Timeout.bound (/usr/src/app/node_modules/generic-pool/lib/ResourceRequest.js:8:15)
    at ontimeout (timers.js:498:11)
    at tryOnTimeout (timers.js:323:5)
    at Timer.listOnTimeout (timers.js:290:5)
```

It was happening with sequelize 4.42.0, mysql2 v1.6.5, node carbon alpine.
After reading a lot about this issue, I figured out that sequelize v5 might handle the connections pool better, so I've tried sequelize 5.7.6, mariadb v2.0.3, but nothing, now the error that I'm getting is:

```
SequelizeConnectionAcquireTimeoutError: Operation timeout
    at pool.acquire.catch.err (/usr/src/app/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:281:53)
    at tryCatcher (/usr/src/app/node_modules/bluebird/js/release/util.js:16:23)
    at /usr/src/app/node_modules/bluebird/js/release/catch_filter.js:17:41
    at tryCatcher (/usr/src/app/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/usr/src/app/node_modules/bluebird/js/release/promise.js:512:31)
    at Promise._settlePromise (/usr/src/app/node_modules/bluebird/js/release/promise.js:569:18)
    at Promise._settlePromise0 (/usr/src/app/node_modules/bluebird/js/release/promise.js:614:10)
    at Promise._settlePromises (/usr/src/app/node_modules/bluebird/js/release/promise.js:690:18)
    at _drainQueueStep (/usr/src/app/node_modules/bluebird/js/release/async.js:138:12)
    at _drainQueue (/usr/src/app/node_modules/bluebird/js/release/async.js:131:9)
    at Async._drainQueues (/usr/src/app/node_modules/bluebird/js/release/async.js:147:5)
    at Immediate.Async.drainQueues (/usr/src/app/node_modules/bluebird/js/release/async.js:17:14)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
```

It's very important to tell that I have different apps that run on the same K8s cluster with the same DB, same auth, and it's not crashing.

Any suggestions?

**Comment by idangozlan** [comment]
My bad, eventually the number of connections was not read correctly from env variables and `acquire` timeout was too low.

Hope it will help anyone else who faces this issue.

**Comment by TrustyMan** [comment]
The same error.

**Comment by TrustyMan** [comment]
{ SequelizeConnectionAcquireTimeoutError: Operation timeout
    at pool.acquire.catch.error (F:\1Daniel\serverless\levar_shopify_integration(completed api)\node_modules\sequelize\lib\dialects\abstract\connection-manager.js:282:52)
  name: 'SequelizeConnectionAcquireTimeoutError',
  parent:
   Error: Operation timeout
       at Timeout._timeout.setTimeout (F:\1Daniel\serverless\levar_shopify_integration(completed api)\node_modules\sequelize-pool\lib\Deferred.js:19:19)
       at ontimeout (timers.js:436:11)
       at tryOnTimeout (timers.js:300:5)
       at listOnTimeout (timers.js:263:5)
       at Timer.processTimers (timers.js:223:10),
  original:
   Error: Operation timeout
       at Timeout._timeout.setTimeout (F:\1Daniel\serverless\levar_shopify_integration(completed api)\node_modules\sequelize-pool\lib\Deferred.js:19:19)
       at ontimeout (timers.js:436:11)
       at tryOnTimeout (timers.js:300:5)
       at listOnTimeout (timers.js:263:5)
       at Timer.processTimers (timers.js:223:10) }

**Comment by iszlailorand** [comment]
@TrustyMan Does a larger ```acquire``` solve your issue?

**Comment by demian85** [comment]
I'm using the latest version and this creepy error still appears randomly. I'm also using the default config values.
The worst part is that my express app completely crashes and stays in a "zombie mode" where I need to reset the server and start from scratch or nothing will ever work again.
Any ideas?

**Comment by papb** [comment]
@demian85 I have no idea to be honest, I really don't know where to start to help you, I know it is probably hard but I think the best course of action for you is to try to find a way to reproduce the problem at will, and then we work from there.

**Comment by jzavisek** [comment]
@demian85 I was seeing this error as well and it seems that in my case it happened when some sequelize calls within a transaction weren't using the parent transaction. So something like:
```js
sequelize.transaction(async transaction => {
  await Model.create({ name: 'Test' }, { transaction })
  await DifferentModel.findAll() // missing transaction
})
```
It seems that these calls blocked the transaction for larger amount of time and other transactions failed with `SequelizeConnectionAcquireTimeoutError`. I am not sure if could be related though.

**Comment by iszlailorand** [comment]
I had this issue as well, solve it by optimizing my queries... I tried all the "hot fixes" and configs that i found without any result, then I tackled the situation and rethink my queries and entities... now everything runs well without any config change.

**Comment by demian85** [comment]
what do you mean by "optimized queries"? what does SQL optimization have to
do with that error?
I'll try to find any transaction issues and report back.

**Comment by iszlailorand** [comment]
@demian85 I couldn't solve this error by any mean. After a lot of investigations I went back to the root of the problem. "Why without any changes this error happens now all the time?" ... end after some research I noticed that I had this error because my volume of data increased. So probably this error is related to that, so I monitorized all my queries and found a very nasty one... I changed it and modify my entities accordingly... now everithing is fine, without any other change.

**Comment by nawlbergs** [comment]
I don't know if this is the same issue.. but sequelize v5 is not workable for us at all at the moment.
Where using `MySql` 
```
 logging:console.log,
 pool: {
     handleDisconnects: true,
     max: 13,
     min: 1, 
     idle: 10000, 
     acquire: 20000 // i also tried 50000
 },
 dialectOptions: {
   ssl: 'Amazon RDS'
},
```
I turned on `logging:console.log` and for some reason now... `Executing (default): SELECT 1+1 AS result` is firing ALL over the place... its happening `at least or two times between EVERY db call`.. and its getting logged `1000+` times eventually..  even tho no database calls are executing.... then it just dies with Operation Timeout and it comes to a halt.
```
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
.... 100's of times this is occuring ...
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
sequelize.js:1187 Executing (default): SELECT 1+1 AS result
SequelizeConnectionAcquireTimeoutError: Operation timeout
    at pool.acquire.catch.error (/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:282:52)
    at <anonymous>

... server is pretty much dead now ...

```
UPDATE:: ... whats extremely odd... is this `SELECT 1+1 AS result` log seems to be tied to the number of results that come back from my queries...  for example... after my user search page... loads (which has a limit of 25)... the `SELECT 1+1 AS result` is logged 25 times... if I search for `nate` and it returns 3 users... it gets logged 3 times at the end...

**Comment by yicheng-irun** [comment]
![image](https://user-images.githubusercontent.com/8521436/71499486-af0b0b80-289b-11ea-9da5-d4deaddf460c.png)
我这里同样是在k8s集群上出现这个问题，我把同一pods下面的另一个协同容器的cpu内存配置改成和主容器一样的配置之后正常了。

**Comment by zabieru96** [comment]
@demian85 You're one of the first persons I've seen that has been experiencing the same issue. Very accurate description.
 
Take a look at this:
https://blog.logrocket.com/why-you-should-avoid-orms-with-examples-in-node-js-e0baab73fa5/

Not saying you should follow the headlines however I found it very insightful. Let me know what you think and if you have resolved the issue. I'll be sure to do the same

**Comment by yicheng-irun** [comment]
> ![image](https://user-images.githubusercontent.com/8521436/71499486-af0b0b80-289b-11ea-9da5-d4deaddf460c.png)
> 我这里同样是在k8s集群上出现这个问题，我把同一pods下面的另一个协同容器的cpu内存配置提升成和主容器一样的配置之后正常了。

提升配置只是延缓了报错的时机。我这个报错的真正原因是因为k8s会发送非常大流量的嗅探 head请求，因为我之前不恰当的使用方式使每一次head请求都会触发一次db的查询，一段时间后，就批量会报这个错误。我后面的处理方式是在express的app注册回调的最前面  app.head('/', (req, res) => { res.send("ok") }); 直接拦截掉head请求，使之不会往下一步的回调走，解决了这个问题。

**Comment by sagitsofan** [comment]
I am also facing this issue, error `SequelizeConnectionAcquireTimeoutError` appears and process is getting stuck, only restart release it.

Any suggestions on how to fix this issue?

**Comment by idangozlan** [comment]
@sagitsofan  It might be that you are running a forked/clustered app, which creates too many connections.
Try to check Max connections config or try to check this aspect too.

**Comment by tapz** [comment]
Started seeing this today. Never seen before.

**Comment by tapz** [comment]
Looks like this started to appear after updating pg to 7.17.0. It has a new option `idle_in_transaction_session_timeout`. Could this be the reason?

**Comment by valentim** [comment]
I get this issue during the development of a task that envolves several inserts (more than 20k) and I solved increasing the `acquire` attribute of the pool option (from 30 seconds to 60 seconds), my configuration below:

`return new Sequelize(
    'DATABASE_URL',
    {
        logging: false,
        pool: {
            max: 30,
            min: 0,
            acquire: 60000,
            idle: 5000
        } 
    }
);`

**Comment by tapz** [comment]
I don't see how 60 seconds instead of 30 seconds would be a solution. At least in my system even 2 seconds is a long time to wait. Besides, downgrading to pg 7.16.0 solved the issue.

**Comment by valentim** [comment]
Neither downgrading nor increasing the timeout settings are ideal solutions, however, both solved the problem, so when someone finds this issue she can choose which way will be better for its proposal...

**Comment by tapz** [comment]
Neither of these are actual solutions. Both are just work arounds. Not sure if the reason for the issue is the same in all cases, but in my case either Sequelize needs some change to work properly with pg 7.17.0 or I need to somehow pass the transaction session timeout value to pg.

**Comment by ezze** [comment]
@tapz, I faced the simlar issue few days ago but was upgrading pg from 7.14.0 to 7.15.1. So 7.17.0 is not an issue in my case for sure. Probably, it's caused by some code refactoring I made earlier, so I have to check it first. Is it possible that some unclosed transaction leads to this error?

**Comment by sagitsofan** [comment]
> this



> I am also facing this issue, error `SequelizeConnectionAcquireTimeoutError` appears and process is getting stuck, only restart fixed it until the next time.
> 
> Node 12.13.0
> Sequelize 5.21.2
> MSSQL
> 
> Any suggestions on how to fix this issue?

So after some investigation i found out that my problem was that one of my queries causes multiple dead lock in my MSSQL db, which eventually raises the `SequelizeConnectionAcquireTimeoutError` error.

I fixed the dead lock issue and `SequelizeConnectionAcquireTimeoutError` error was disappear.

**Comment by tapz** [comment]
@ezze Well, in my case it started when I deployed with 7.17.0 and disappeared when I deployed again with 7.16.0 and no other changes.

**Comment by ezze** [comment]
I found at least one place in my code that could lead to this error. Anyone facing the same issue be sure that you commit or rollback all transactions first. In my case I forgot to commit a transaction in one of my `if` branches.

```javascript
const transaction = await connection.transaction();
try {
  task = await resolveTask(key, { excludeMetadata, transaction, withProducts: true });
  if (task === null) {
    await transaction.commit();
    return new ApiResponse(API_RESPONSE_CODE_NOT_FOUND, 'Task is not found.');
  }
  if (task.status !== STATUS_PENDING && (task.status !== STATUS_DONE || task.aggregatedJobStatus !== STATUS_PAUSED)) {
    return Promise.reject('Task must be pending or done with paused jobs.');
  }
  task.priority = parsePriority(priority);
  await task.save({ transaction });
  await transaction.commit();
}
catch (e) {
  await transaction.rollback();
  if (typeof e === 'string') {
    return new ApiResponse(API_RESPONSE_CODE_INVALID, e);
  }
  throw e;
}
```

In order check whether some PostgreSQL transactions are idle one could execute the following SQL-query (`geobase` is database name in the example):

```sql
select * from pg_stat_activity where datname = 'geobase' and state = 'idle in transaction';
```

**Comment by ezze** [comment]
@jzavisek, finally realized that you were right. A forgot to pass my transaction to inner function and transaction with `default` identifier was created instead. In result my transaction was idle for 60 seconds and in final specified `SequelizeConnectionAcquireTimeoutError` was thrown. Now everything works for me, even after upgrade to `pg@7.17.1`.

**Comment by kennetpostigo** [comment]
Ran into this issue as well, going to try downgrading to `7.16.0`. Is the sequelize team aware of this issue with newer versions of `pg`?

**Comment by ebaynaud** [comment]
Do you guys use RDS in multi-AZ mode?

**Comment by kumanoko24** [comment]
After hours of debugging, I found a pattern which is not easy to be found when using manual transaction commit and rollback in async/await:

```javascript
const dbtx = await db.transaction();

if (_records_materials.length === 0) {
  return true; // but no commit or rollback yet
}

await dbtx.commit()
```

**Comment by royalpinto** [comment]
> It seems that these calls blocked the transaction for larger amount of time and other transactions failed with `SequelizeConnectionAcquireTimeoutError`. I am not sure if this could be related though.



> @demian85 I was seeing this error as well and it seems that in my case it happened when some sequelize calls within a transaction weren't using the parent transaction. So something like:
> 
> ```js
> sequelize.transaction(async transaction => {
>   await Model.create({ name: 'Test' }, { transaction })
>   await DifferentModel.findAll() // missing transaction
> })
> ```
> 
> It seems that these calls blocked the transaction for larger amount of time and other transactions failed with `SequelizeConnectionAcquireTimeoutError`. I am not sure if this could be related though.

Created a bug for this: https://github.com/sequelize/sequelize/issues/11995

**Comment by OnArs** [comment]
> I don't see how waiting 60 seconds instead of 30 seconds would be a solution. At least in my system even 2 seconds is a long time to wait. Besides, downgrading to pg 7.16.0 solved the issue.

Sorry, but what is pg?

**Comment by marcogrcr** [comment]
Keep in mind that if you use `new Sequelize({ pool: { max: 1 }, /* (...) */ })` (common config for AWS Lambda functions) and use `Model.findAndCountAll()`, you may experience `SequelizeConnectionAcquireTimeoutError` errors for expensive (latency-wise) queries.

See #12633 for more details.

**Comment by xuqingkuang** [comment]
Same issue here, solved by restart the server temporary, but need a solution.

Config:
```typescript
const config: Options = {
  pool: {
    max: 5,
    min: 0,
    acquire: 60000,
    idle: 60000,
  },
};
```

![log](https://user-images.githubusercontent.com/32739/95724063-7b716480-0ca8-11eb-98dc-9149cf965e1b.png)

**Comment by mahiraltinkaya** [comment]
I got this problem every time I was inserting 7 thousand rows of data. For this, I overcame the problem by using for of and await method each time and also by setting the mysql max connection to 5000. I am not facing any problem right now.

**Comment by RNm-dove** [comment]
I had the same error and found the reason.
In my case,  this is because sequelize transaction consume the number of connection pools depending on the number of models used in the transaction scope.

For example, I set pool.max to 2 in sequelize configure. I use two sequelize model in one transaction scope.
```
const sequelize = new Sequelize(
  config.database,
  config.username,
  config.password,
  {
    host: config.host,
    dialect: config.dialect
    logging: config.logging,
    pool: {
      max: 2 // connection pools max 2
   }
  },
);

class FooService {
  static async fooServiceFunction(){
  const transaction = await sequelize.transaction();
  
  const firstModel =  await SequelizeFirstModel.findByPk(1, {
    transaction,
    lock: transaction.LOCK.UPDATE,
  });
  
  // do something
  const second = {
    someKey: firstModel.someValue
  }
  
  await SequelizeSecondModel.create(second);
  
  await transaction.commit();
  }
}

```

Then, call fooService at the same time. You will get connection timeout error.
```
FooService.fooServiceFunction();
FooService.fooServiceFunction();

// output sequelize log
Executing (dc2376d7-0198-412e-8a64-81e9b8718031): START TRANSACTION;
Executing (85f89fe9-e896-419d-88ef-3d6016ba5534): START TRANSACTION;
Executing (dc2376d7-0198-412e-8a64-81e9b8718031): SELECT "id", "someValue"  FROM "SequelizeFirstModel" WHERE "SequelizeFirstModel"."id" = 1 FOR UPDATE;
Executing (85f89fe9-e896-419d-88ef-3d6016ba5534): SELECT "id", "someValue"  FROM "SequelizeFirstModel" WHERE "SequelizeFirstModel"."id" = 1 FOR UPDATE;
```

But, this work when max pool connection is 3
```
const sequelize = new Sequelize(
  config.database,
  config.username,
  config.password,
  {
    host: config.host,
    dialect: config.dialect
    logging: config.logging,
    pool: {
      max: 3 // connection pools max 3
   }
  },
);
```

Maybe, each transactions wait until they can have two connection pool, so connection time out occur.

**Comment by royalpinto** [comment]
@RNm-dove Seems like it is because you are not using transaction for `await SequelizeSecondModel.create(second);`.
When not using a transaction, it will try to make a new connection but that connection is already acquired by a second call to `FooService.fooServiceFunction()`.

`await SequelizeSecondModel.create(second, { transaction })` should solve the problem.

**Comment by RNm-dove** [comment]
@royalpinto 
I understand ! Thunk you very much 😊

**Comment by arbieo** [comment]
Since this is the top search result on google: sequelize has a way to automatically pass transactions to all your calls documented here:

https://sequelize.org/master/manual/transactions.html#automatically-pass-transactions-to-all-queries

So you can make your code shoot-yourself-in-the-foot proof

**Comment by achempak** [comment]
I'm sure there are multiple causes for this error. For me, I was accumulating thousands of transaction promises in a loop and executing them all at once with Promise.all. This worked for most use cases, but there were times (when the number of transactions was 80,000+) where the app would crash and this would be the error message. I'm assuming its a memory issue, but that was impossible to tell given the message. Once I switched to performing the transactions sequentially in the loop, the issue was solved (albeit, the whole process was significantly slower). For better performance, I probably could've batched the transaction promises in groups of ~20,000. For my particular use-case, however, performance wasn't a big issue.

**Comment by amonemian** [comment]
I noticed that **awaiting** for the transaction's Commit() and Rollback() solved my issue. As in my case, the commit and rollback return a promise as well. So it makes the transaction wait for the results of the queries and prevents the inconsistencies.

**Comment by lardcanoe** [comment]
I was upgrading node from v12 to v14 and it was hanging during the database connection. For me, the issue was that our package.json locked pg at v7.16. When I updated pg to v8.7.1 everything worked perfectly.

**Comment by rick-seamless** [comment]
This will probably help no one, but my issue was that my Aurora read replica was falling too far behind the writer and was restarting

**Comment by Jamie0** [comment]
This issue had been plaguing us for a while (it was happening often on launch). After firing the app up with a debugger, I noticed a 10 second idle period following a DNS resolution.

It turns out that Node's built-in DNS module will only handle 1 DNS lookup at a time, and queues the rest. 

Combining this bottleneck with an I/O bottleneck of modules instantiating (require-ing a lot of files), it often took 10+ seconds for the DNS callback to fire, resulting in this timeout. This was exacerbated by the connection pool trying to open some sockets simulatneously. 

Changing localhost to 127.0.0.1 (or otherwise using an IP address) completely cured this problem for me.

**Comment by PythonCoderAS** [comment]
> This issue had been plaguing us for a while (it was happening often on launch). After firing the app up with a debugger, I noticed a 10 second idle period following a DNS resolution.
> 
> It turns out that Node's built-in DNS module will only handle 1 DNS lookup at a time, and queues the rest.
> 
> Combining this bottleneck with an I/O bottleneck of modules instantiating (require-ing a lot of files), it often took 10+ seconds for the DNS callback to fire, resulting in this timeout. This was exacerbated by the connection pool trying to open some sockets simulatneously.
> 
> Changing localhost to 127.0.0.1 (or otherwise using an IP address) completely cured this problem for me.

I think this is one of the causes. When the internet for my server failed, DNS lookups to external addresses would hang indefinitely. I suspect that when the pool tried to make a new correction during this hanging, it would timeout and crash.

**Comment by 7hibault** [comment]
@dijonkitchen I think I had the same issue. I could stop having this error with one of the follow 3 methods : 
- increasing the pool configuration (allowing only 1 connection makes it break, allowing 100 makes it pass, or increasing acquire time makes it work with some waiting time)
- getting the queries out of a transaction
- unparalleling and processing the queries one by one 

None of those seem entirely satisfying
- increasing the config may allow the break to happen again
- getting the query out of a transaction makes us lose the benefits of using a transaction in the first place
- unparalleling makes things go slower

I don't understand why it works outside a transaction. If it's only a matter of concurrent connections, what does it change that the queries are inside a transaction ?

**Comment by dijonkitchen** [comment]
The fix was using a single transaction for **all** of them, rather than a transaction for **each** of them

**Comment by 7hibault** [comment]
@dijonkitchen Does it mean that transactions use a separate connection each ? Because getting the queries out of a transaction also worked to avoid the timeout.

**Comment by ephys** [comment]
Transactions are always run on their own connection, there can only be one transaction per connection (by design in sql) :)

**Comment by arbieo** [comment]
And to reiterate

Since I know many people reading this thread are going to try and manually track the transaction object to pass it to each query. This would indeed be a correct solution however it's error prone (especially in nested functions) and forgetting it in a single place can deadlock your database connections. (If you are holding a transaction connection, max your db connections and then try to open a new connection)

Please see here to learn how to add transactions to all queries inside a transaction block automatically: https://sequelize.org/master/manual/transactions.html#automatically-pass-transactions-to-all-queries

**Comment by fanmaomao** [comment]
I had the same `SequelizeConnectionAcquireTimeoutError` and write an example to reproduce it. 
May be we need use `transaction` in `sequelize.transaction(async transaction => { ... })` all the time, but sometimes ...

![screenshot-20220516-113159](https://user-images.githubusercontent.com/13829872/168516523-74fa465b-12f7-43ad-a04a-9636e92968bb.png)
![image](https://user-images.githubusercontent.com/13829872/168516590-158be145-9193-47a5-9bd2-aef3dedfe605.png)

```javascript
const Sequelize = require("sequelize");

async function main() {
    const sequelize = new Sequelize({
        dialect: "mysql",
        timezone: '+08:00',
        logging: console.log,

        database: "",
        host: "",
        port: "",
        username: "",
        password: "",

        pool: {
            max: 3,
            min: 0,
            acquire: 1000 * 10,
            idle: 1000 * 30,
        }
    });
    const pool = sequelize.connectionManager.pool;
    // console.log(pool);
    // return;

    let i = 0;
    setInterval(() => console.log("timer", ++i), 1000);

    await Promise.all(new Array(5).fill().map(() => sequelize.transaction(async transaction => {
        await sequelize.query("SELECT 'transaction';", { type: Sequelize.QueryTypes.SELECT, transaction });
        console.log(`connection pool using:${pool.using} available:${pool.available}`);
        // await sequelize.query("SELECT 'normal';", { type: Sequelize.QueryTypes.SELECT, transaction });
        await sequelize.query("SELECT 'normal';", { type: Sequelize.QueryTypes.SELECT });
    })));

}

main().catch(err => console.log(err));
```

**Comment by ephys** [comment]
@arbieo I think we should turn that on by default once we replace cls with node's built-in `AsyncLocalStorage`.

**Comment by arbieo** [comment]
@ephys that would be lovely and probably enough to mark this closed when done. Though I do worry about backwards compatibility if it's turned on for everyone. Would need to be a major version at minimum.

**Comment by ephys** [comment]
I'm not convinced it's the whole story. I fear [`sequelize.transaction`](https://github.com/sequelize/sequelize/blob/main/src/sequelize.js#L997-L1018) may, in some scenarios, result in a connection being left with an uncommitted transaction. This may eventually lead all other connections to wait for that transaction to be committed, leading to the whole pool being in use, and a `SequelizeConnectionAcquireTimeoutError`.

**Comment by ephys** [comment]
The patch has been released https://github.com/sequelize/sequelize/releases/tag/v6.20.1
Please let me know if you're still experiencing the issue with this change :)

**Comment by AliHewiagh** [comment]
Please make sure to roll back/commit your transaction. My case was initiating the transaction and in certain flow, I return a response without rolling back or committing my transaction.

**Comment by canove** [comment]
> The patch has been released https://github.com/sequelize/sequelize/releases/tag/v6.20.1 Please let me know if you're still experiencing the issue with this change :)

I've applied this patch and not more timeout until now (7 days and conting)

thanks!

**Comment by WikiRik** [comment]
I'll close this issue since it seems to be completed. If people still come across this issue they can feel free to comment here

**Comment by MathiasMoralesDixtra** [comment]
I continue with this problem after the update.
web: {"level":"error","message":"meliAddShipping Operation timeout","name":"SequelizeConnectionAcquireTimeoutError","original":{},"parent":{},"stack":"SequelizeConnectionAcquireTimeoutError: Operation timeout\n    at ConnectionManager.getConnection (/var/app/current/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:204:15)\n    at async Transaction.prepareEnvironment (/var/app/current/node_modules/sequelize/lib/transaction.js:85:24)\n    at async /var/app/current/node_modules/sequelize/lib/sequelize.js:500:7\n

**Comment by ephys** [comment]
We're still monitoring this thread

`SequelizeConnectionAcquireTimeoutError` means all connections currently in the pool were in-use and a request to access a connection took more than 60 seconds (default timeout).

This can typically happen if there are too many requests being made and the pool is too small.

If the pool size is not the issue, unfortunately the stacktrace is not useful to us as it only lets us know that the pool is saturated, not why all connections are in-use.

Consider taking a look at what the connections are doing (if using postgres, pgadmin has a dashboard that displays that nicely). If they're not doing anything (ilde), but sequelize's pool says they are all in-use, then we likely have a bug in Sequelize.

If that is the case, please let us know which version of Sequelize you're using, as well as the name of the database, its version, and the version of the npm library to use to connect to it

**Comment by saqib-ahmed** [comment]
Thanks for your response @ephys. I'll have a look at pgAdmin and let you know what I find.

**Comment by saqib-ahmed** [comment]
I found my issue. I was starting a transaction at the top of my function which was set to execute in loop. There was one case where I wasn't commiting/rollbacking the transaction. Fixing that resolved the issue.

**Comment by DewangS** [comment]
Hi @saqib-ahmed 
I am having similar issue though against MariaDB (MySQL) hopefully this is not specific to the database. In terms of transaction management, I'm using managed transactions so, no manual commit\rollback for me ([Sequlize documentation](https://sequelize.org/docs/v6/other-topics/transactions/)). What is your pool config as I am using the default i.e. I don't have pool{} segment in my config. In my use case, I process 100-200 records in a loop.

**Comment by saqib-ahmed** [comment]
Hi @DewangS 

I'm also using the default pooling config with PostgreSQL, but I'm managing my transactions manually. If you're letting Sequelize handle the transactions, it might be due to a bug in Sequelize. 
A good starting point for troubleshooting the issue would be switching to manual transaction handling and seeing if the issue still occurs.

**Comment by saqib-ahmed** [comment]
Hi @DewangS 

I'm also using the default pooling config with PostgreSQL, but I'm managing my transactions manually. If you're letting Sequelize handle the transactions, it might be due to a bug in Sequelize. 
A good starting point for troubleshooting the issue would be switching to manual transaction handling and seeing if the issue still occurs.

**Comment by DewangS** [comment]
Thanks @saqib-ahmed for the tips though the same code was working fine with MySQL but since we moved to MariaDB, I am seeing these errors. Also, as the project has been live for more than a year with managed transactions, it would be a BIG change to update it to manual transactions. Will require a lot of time to re-test everything. I'll see if the pool settings helps.

**Comment by DewangS** [comment]
Hi @saqib-ahmed ok, we have found the solution while it works, I am not convinced. so, we increased the connection setting on MariaDB as per below which I believe is excessively high

[mariadb]

max_user_connections=400
max_connections=800

in my connection pool settings I have below ...once again I've a very high max connections and high connect and acquire timeout values.

"test": {
    "database": "permits",
    "host": "DB hostname",
    "port": 3306,
    "dialect": "mysql",
    "dialectOptions": {
       **"connectTimeout": 220000**
    },
    "pool": {
      **"max": 350**,
      "min": 0,
      **"acquire": 220000**,
      "idle": 10000
    },
    "define": {
      "freezeTableName": true,
      "timestamps": false
    }
  },

I could manage to process a file with more than 400 records though I would like to know if each insert\update which In perform within the loop would need a separate DB connection? if so, that's unacceptable.  Is it because Sequelize ORM works on promises and each promise will hold on to a connection till it finishes? it seems more like the case. if so, how do you throttle promises so that they don't get fire in a one major burst?

**Comment by PythonCoderAS** [comment]
@DewangS you could probably use a queue. I wrote my own package to queue up promises as an example: https://www.npmjs.com/package/concurrent-priority-worker-queue

**Comment by DewangS** [comment]
Thanks @PythonCoderAS Queue was one of the option though I really don't like these promises esp. coming from pure Java background. This seems to be design flaw considering node is a single threaded. Literally one should avoid node based backend for any heavy lifting. May be good for simple REST APIs. I was wondering if you can use settInterval etc in between to put some pause between promises? any idea?

**Comment by igorchvalun** [comment]
How I solved the issue
Background. I was parsing some Excel files. Each parsed row was supposed to be parsed into few entities and stored to DB. With some smaller files I had no issue. 8mb file caused the sequelize error and crash. Process: parse a row -> callback -> map to internal object 1 -> store -> map to internal object 2 -> store.
What happened according to logs: parse a row -> callback -> parse a row -> callback ... until all rows are parsed. And only then internal object 1 -> store, and then again internal object 2 -> store.
The reason that happened was the parsing library. It used "foreach" loop under the hood. Which caused the error.
My solution: never use "foreach", even if it wasn't you.

**Comment by PythonCoderAS** [comment]
> Thanks @PythonCoderAS Queue was one of the option though I really don't like these promises esp. coming from pure Java background. This seems to be design flaw considering node is a single threaded. Literally one should avoid node based backend for any heavy lifting. May be good for simple REST APIs. I was wondering if you can use settInterval etc in between to put some pause between promises? any idea?

I didn't see this question at all but the solution is to await a sleep promise. https://stackoverflow.com/a/39914235/12248328

**Comment by fzn0x** [comment]
Since we are still monitoring this issue, I think I will reopen the issue.

**Comment by ephys** [comment]
Reopening indicates that there is something that needs to be fixed on our end. I don't have any evidence of that currently.

- If you use unmanaged transactions, you're fully responsible for making sure they're properly committed / rolled back. Managed transactions are much safer
- If you run too many connections concurrently (such as by looping and running a query for 8mb worth of data), you're going to run out of available transactions for long enough to have a timeout. You need to batch your operations as much as possible, chunk the data, or queue your operations

All new comments I've seen so far fall in either of these two categories, which isn't caused by Sequelize.

> I would like to know if each insert\update which In perform within the loop would need a separate DB connection? if so, that's unacceptable.

If you don't await them then yes it's going to run all of these operations in parallel which is going to require a new connection for each of them. If you await each operation, it's going to run in series. You want the first one for performance. If your number of operations is too big you have a few options:

- increase the maximum number of concurrent connections sequelize can open, and your DB can accept
- use batch queries when possible to reduce the number of necessary queries
- process your data in chunks instead of all at once
- use a queue which only processes a few entries in parallel at a time

You'd have the same issue in Java if you tried to parallelize this work. If you don't, then it's going to run in series and will be very slow. It's the same thing, just that Node makes it easier to parallelize by default.

**Comment by lxcid** [comment]
If we enable CLS, is it possible to add a hook to report when the CLS automatically pass the transaction to our query?

This can help us catch coding bugs.

**Comment by ephys** [comment]
CLS passes the transaction to all queries run by a transaction's callback. You can use the `beforeQuery` hook to check whether a transaction is used:

```ts
sequelize.beforeQuery((options, query) => {
  console.log(options.transaction);
});
```

**Comment by andrewinsoul** [comment]
check your code carefully - there must be an opened transaction that was not either committed or rolled back thereby leaving the transaction hanging. 
Most times, developers open a transaction on the first line of the function and put transaction.commit() on the last line of the function and transaction.rollback() in the catch block but some checks are done in that function and sometimes, the function is returned at some point before getting to the transaction.commit line. What I just described here is a very common culprit of the error you just described. A snippet of the code will aid in debugging the issue

**Comment by reunix** [comment]
Hello everyone, as it is not very old and I am having the same problem, I come to share how I solved it in a way.

Changing the parameters of idle, acquire, max, min .....
nothing solved, I managed to install the mysql package and working together,

the queries before the transactions, I do with mysql directly, and the transactions continue with sequelize, thus using both.

what happens here when there is a large amount of transaction requests, the queries that precede leave the sequelize service blocked, as if it had several open connections, it was like this now:

**for queries:**
` const pool = mysql.createPool(dbconfig.connectionMultipleStatements)
    return new Promise((resolve, reject) => {
        pool.query(sql, (error, elements) => {
            pool.end()
            if (error) {
                console.log('Erro: VerificarEstoquePontoEntrega: ', error)
                return reject(error);
            }
            resolve(elements);
        });
    });`

**for transaction:**
`await sequelize.query(sql, { type: sequelize.QueryTypes.UPDATE, transaction: transactionAddItem })`

**Config sequelize:**
`  pool: {
            // match: [/Deadlock/i, Sequelize.ConnectionError],
            max: 5,
            min: 0,
            idle: 5_000, // 10_000
            acquire: 10_000, // 60
        }`

**a tip is to enable the**
`logging: console.log,`

to make sure all TRANSACTIONS contain a ROLLBACK or COMMIT

bye!

**Comment by jamesicon** [comment]
We are also running into this issue, if a certain endpoint that creates a transaction in the beginning is called in quick succession it eventually locks the entire system up. I have increased the pool size to help with this but it seems like there should be a way to have transactions queue themselves vs stopping everything completely once the connection pool is used up

**Comment by weisinlakd** [comment]
Would it be possible to add a suggestion about the possible problems/solutions on the error log report? Or even a link to this issue? 
I think it may solve a few headaches, especially for new developers or people that are starting to use transactions, since the error itself is kind of a headscratcher because it "shows up out of nowhere" and is hard to replicate IF you dont know what is actually happening (pools, transactions, etc)

**Comment by nabil-hamisa** [comment]
The problem I encountered was related to the transaction, it was initiated but neither committed nor rolled back. Here's an example of the code where this issue occurred:
`router.post("/api/dosomething", async (req, reply) => {
  // Incorrect transaction placement
  const transaction = await sequelize.transaction();
  const user = User.findOne({ where: { id: user.id } });
  if (!user) {
    await transaction.rollback();
    return reply.code(400).send('User not found');
  }
// =====>here transaction should be place  <======
  try {
    // Business logic
    await transaction.commit();
  } catch (e) {
    console.log(e);
    await transaction.rollback();
    return reply.code(500);
  }
    return reply.send({ message: "Add something successful" });

});
`

**Comment by ephys** [comment]
ConnectionAcquireTimeout errors are notoriously difficult to debug because they are usually a symptom of various problems that are difficult to pin down

The documentation has a section about the error that can help point you in the right direction. It's written for v7, but is relevant to v6 as well: https://sequelize.org/docs/v7/other-topics/connection-pool/#connectionacquiretimeouterror

**Comment by moonjoungyoung** [comment]
@ephys 

We're experiencing a SequelizeConnectionAcquireTimeoutError: Operation timeout in our application. The issue is occurring only on some of our servers, while others are functioning normally. This indicates that the problem is not with the database itself, but rather server-side.

We suspect there might be unclosed Sequelize transactions, but our app is too large to pinpoint the exact location. To investigate, we've implemented logging for the connection pool before and after each request:
const { pool } = sequelize.connectionManager;
return pool: ${req.id} ${pathname} ${type} | total: ${pool.size}, using: ${pool.using}, waiting: ${pool.waiting}, available: ${pool.available};
Our pool configuration:

Size: 0-288 per server
Acquire: 30000
Idle: 10000

We have 8 to 24 servers with auto-scaling, and our database can handle over 5000 connections.
Oddly, our logs show something like this:
total: 17, using: 16, waiting: 0, available: 1
This indicates that the pool has available connections, yet we're still getting timeout errors.
Once the error occurs, it affects all subsequent API requests on that particular server. We typically resolve this by restarting the affected server, but the error reoccurs after 1-2 days.
Additional details:

Sequelize version: 6.37.3
The issue is isolated to specific servers; other servers continue to operate without problems, which further confirms that this is not a database-level issue.

We're seeking assistance in resolving this persistent issue and understanding why we're getting timeout errors on some servers despite available connections in the pool. Any insights or suggestions you could provide would be greatly appreciated.
Thank you for your time and expertise. We look forward to your response.

**Comment by a714363021** [comment]
In v5 version, As long as the concurrency exceeds pool.max, all requests will be reported as errors.

**Comment by TheMacros** [comment]
> [@ephys](https://github.com/ephys)
> 
> We're experiencing a SequelizeConnectionAcquireTimeoutError: Operation timeout in our application. The issue is occurring only on some of our servers, while others are functioning normally. This indicates that the problem is not with the database itself, but rather server-side.
> 
> We suspect there might be unclosed Sequelize transactions, but our app is too large to pinpoint the exact location. To investigate, we've implemented logging for the connection pool before and after each request: const { pool } = sequelize.connectionManager; return pool: ${req.id} ${pathname} ${type} | total: ${pool.size}, using: ${pool.using}, waiting: ${pool.waiting}, available: ${pool.available}; Our pool configuration:
> 
> Size: 0-288 per server Acquire: 30000 Idle: 10000
> 
> We have 8 to 24 servers with auto-scaling, and our database can handle over 5000 connections. Oddly, our logs show something like this: total: 17, using: 16, waiting: 0, available: 1 This indicates that the pool has available connections, yet we're still getting timeout errors. Once the error occurs, it affects all subsequent API requests on that particular server. We typically resolve this by restarting the affected server, but the error reoccurs after 1-2 days. Additional details:
> 
> Sequelize version: 6.37.3 The issue is isolated to specific servers; other servers continue to operate without problems, which further confirms that this is not a database-level issue.
> 
> We're seeking assistance in resolving this persistent issue and understanding why we're getting timeout errors on some servers despite available connections in the pool. Any insights or suggestions you could provide would be greatly appreciated. Thank you for your time and expertise. We look forward to your response.

Did you find the solution?

**Comment by vvictorfonseca** [comment]
In my case, the root cause was a problematic query with a wrong set of include options in Sequelize.
Because of that, the query was taking an extremely long time to resolve. Under normal usage it was already slow, but with higher traffic the connections started to pile up in the pool, and eventually Sequelize hit the SequelizeConnectionAcquireTimeoutError.

So the issue was not with Sequelize itself, but with how the query was structured — the long-running query was blocking the pool, and other requests timed out waiting for a free connection.
